scalyr_agent/sharded_copying_manager/copying_manager.py

```diff
@@ -1256,4 +1202,168 @@ def __scan_for_new_logs_if_necessary(
+            m = re.search(
+                r".*/checkpoints-(?P<id>\d+_\d+)\.json", full_checkpoints_path
+            )
+            if m:
+                worker_id = m.group("id")
+                active_checkpoints_path = os.path.join(
+                    checkpoints_dir_path, "active-checkpoints-%s.json" % worker_id
+                )
+            else:
+                active_checkpoints_path = None
+
+            checkpoints = self.__read_worker_checkpoint_state(
+                full_checkpoints_path, active_checkpoints_path, current_time
+            )
+
+            if checkpoints:
+                workers_checkpoints.append(checkpoints)
+
+        return workers_checkpoints
+
+    def __read_worker_checkpoint_state(
+        self, full_checkpoint_file_path, active_checkpoint_file_path, current_time
+    ):
+        """Reads the checkpoint state from disk and returns it.
+
+        The checkpoint state maps each file path to the offset within that log file where we left off copying it.
+
+        @return:  The checkpoint state
+        @rtype: dict
+        """
+
+        if not os.path.isfile(full_checkpoint_file_path):
+            log.info(
+                'The log copying checkpoint file "%s" does not exist, skipping.'
+                % full_checkpoint_file_path
+            )
+            return None
+
+        # noinspection PyBroadException
+        try:
+            full_checkpoints = scalyr_util.read_file_as_json(
+                full_checkpoint_file_path, strict_utf8=True
+            )
+
+            if active_checkpoint_file_path is not None and os.path.isfile(
+                active_checkpoint_file_path
+            ):
+
+                # if the active checkpoint file is newer, overwrite any checkpoint values with the
+                # updated full checkpoint
+                active_checkpoints = scalyr_util.read_file_as_json(
+                    active_checkpoint_file_path, strict_utf8=True
+                )
+
+                if active_checkpoints["time"] > full_checkpoints["time"]:
+                    full_checkpoints["time"] = active_checkpoints["time"]
+                    for path, checkpoint in six.iteritems(
+                        active_checkpoints["checkpoints"]
+                    ):
+                        full_checkpoints[path] = checkpoint
+
+            if current_time - full_checkpoints["time"] > self.__config.max_allowed_checkpoint_age:
+                log.warn(
+                    "The worker checkpoint file '%s' is too stale (written at '%s').  Ignoring it.  All log files will "
+                    "be copied starting at their current end.",
+                    full_checkpoint_file_path,
+                    scalyr_util.format_time(full_checkpoints["time"]),
+                    error_code="staleCheckpointFile",
+                )
+
+                # checkpoint files are stale, so remove them.
+                os.unlink(full_checkpoint_file_path)
```

Comment:
This seems unsafe because it looks like we are removing the checkpoint state from disk.  What happens if the agent crashes before it is able to write it to disk?  We'd lose all of the checkpoint state.

I think we need to come up with a better strategy on how to handle start up and consolidating the checkpoint state.

My suggestion is that we have an initialization step that reads all checkpoints (including legacy file paths) and writes out a new, single consolidated file to disk before it goes and deletes all of the other checkpoint files.  Once we have cleaned up the state like that, then we can have the individual workers start working, writing out their own checkpoint files, etc.

Reply:
As discussed, now the copying manager combines all checkpoint files into a single - master file. After that, it also removes the worker checkpoints.